### ISSUE TITLE
Exception in msvc2013

### ISSUE THREAD

**Issue by cstamatopoulos** [KjellKod/g3log#61]
I am not sure why this happens but the code below throws an exception in msvc2013 but it is all good in 2015. The exception is thrown at the lock in shutDownLogging.

    auto worker = g3::LogWorker::createLogWorker();

    int main(int argc, char* argv[])
    {
    	using namespace g3;
    	auto defaultHandler = worker->addDefaultLogger(argv[0], "./");
    	g3::initializeLogging(worker.get());
    }

Now you might ask why someone would want to do that but unfortunately for old apps with MFC there is no main function, just a WinMain which you dont have access to.

**Comment by KjellKod** [comment]
What it the exception? (Given by std::exception::what() function)

**Comment by cstamatopoulos** [comment]
It is an Unhandled exception.
I thought it was related to shutDownLogging so I moved it from internal to g3 namespace to call it manually. While there was no exception at the call, I did get the same exception later on when shutDownLogging was called in the destructor of ~LogWorker

**Comment by KjellKod** [comment]
if I understand you correctly you create and keep the ```worker``` in an anonymous namespace (or just global in the cpp file), right? 

I've seen on some systems, with some compilers that destruction of unique_ptrs can go wrong if it is kept as a static variable (or in anonymous namespace, or just globally in a cpp file). 

There are a couple of workarounds. One that worked for at least one compiler (don't remember which) was to change the container from a unique_ptr to a shared_ptr. Please try that and get back to me.

i.e. 
```
namespace {
   auto worker = std::shared_ptr<g3::LogWorker>(g3::LogWorker::createLogWorker().release());
}

int main(int argc, char* argv[])
{
    using namespace g3;
    auto defaultHandler = worker->addDefaultLogger(argv[0], "./");
    g3::initializeLogging(worker.get());
}
```

**Comment by KjellKod** [comment]
please try the above.   I will also close this soon unless it can be shown that the exception is caused by a bug in g3log. it seems that it is more of a case of how the initialization is done and the quirkiness of that specific compiler. 

(even though I might close this soon, I will continue to work with you to help you out... )

**Comment by cstamatopoulos** [comment]
Yes you understand correctly.
It is always in the LogWorker destructor that the crash happens.
Using a shared_ptr does results in the same thing.

The only thing that I could do to make it work is create the unique_ptr and release it to a raw ptr.
Then manually call shutdown and  then delete the raw ptr. But that is not really clean while I also had to move functions from internal namespace. Btw the doc says that you can manually call shutdown but since then its been probably moved to internal.

**Comment by KjellKod** [comment]
I don't have access to a system like yours so I cannot test t but I remember that there was another developer with this issue. 

In the end I think it was enough to make he logworker instance static in **function** scope, not global scope

See here for details: 
https://github.com/KjellKod/g3log/issues/35

**Comment by cstamatopoulos** [comment]
program never terminates with static in function scope, it just waits on token_cleared.wait();

**Comment by KjellKod** [comment]
OK. I will try to spin up a msvc2013 during the holidays and see if I can reproduce this. 

Can you please send me exactly a functional example minimized to a file or two that exhibits this issue?
You should be able to attach a small zip file to this issue thread.  I.e. make it so that I only have to unzip it and open it in visual studio to get it to work.

**Comment by cstamatopoulos** [comment]
I've created a full solution for you so you just have to open the .sln file and run it without the need of setting up anything in msvc. The g3log current master has been added to the solution.
Include/ input libs have only been set for the x64 configurations.
The debug version will crash while the release version hangs.

Link added to my bod as you cannot added anything but basic files here.
https://app.box.com/s/b77cfvwu6b682ihcfiai6424k8y53y20

If you just want the main.cpp I can provide that it really is just a few lines

**Comment by KjellKod** [comment]
I changed to code to what it should look like for a normal project
```
int main(int argc, char* argv[])
{
	using namespace g3;
	auto worker = g3::LogWorker::createLogWorker();
	auto defaultHandler = worker->addDefaultLogger(argv[0], "./");

	// logger is initialized
	g3::initializeLogging(worker.get());

#if _DEBUG
	std::cout << "Will crash";
#else
	std::cout << "Will hang";
#endif
	return 0;
}
```

Both Debug and Release worked great.   I guess the issue might be in the WinMain or lack thereof that you are talking about. In the example you provided I cannot see the issue. 

<img width="489" alt="screen shot 2015-11-24 at 10 52 07 pm" src="https://cloud.githubusercontent.com/assets/331742/11389477/64251318-92fe-11e5-8a74-cd73be1dd34d.png">
<img width="533" alt="screen shot 2015-11-24 at 10 51 32 pm" src="https://cloud.githubusercontent.com/assets/331742/11389478/64255d82-92fe-11e5-8ca5-bc39ddf591b1.png">

**Comment by cstamatopoulos** [comment]
I know that, but can you see the same behavior if you keep the code as I sent it? There is something funny going on when worker is global, and I am not saying a bug on your side. Could be the compiler as msvc2015 is fine.

**Comment by KjellKod** [comment]
Did you see this?   https://github.com/KjellKod/g3log/issues/35

**Comment by cstamatopoulos** [comment]
Yes making it a static made it hang debug.I posted it above in a comment.

**Comment by KjellKod** [comment]
Do you know when your process is exiting?

I.e. would it be possible to call something then?

**Comment by cstamatopoulos** [comment]
Yes the exiting part is OK, there is an exitinstance function in mfc.

I had a workaround using that. Create the unique ptr then release it. On
exit manually call shutdown then delete the raw ptr. I think that worked
(would have to test it again) but i had to move the shutdown out of the
internal namespace.

**Comment by KjellKod** [comment]
no. don't touch the shutdown. Like I said earlier it doesn't work that way.  You need a clean RAII shutdown. 

I have an example below that I think you can utilize. 
It shows how you can retrieve a shared_ptr<LogWorker> and in your exit function you can reset that pointer. 


```
#include <iostream>
#include <g3log/g3log.hpp>
#include <g3log/logworker.hpp>
#include <g3log/std2_make_unique.hpp>

std::shared_ptr<g3::LogWorker> init(std::string prefix)
{
	 std::shared_ptr<g3::LogWorker> gMainLogger(g3::LogWorker::createLogWorker().release());

	auto defaultHandler = gMainLogger->addSink(std2::make_unique<g3::FileSink>(prefix, "./"),
		&g3::FileSink::fileWrite);
	g3::initializeLogging(gMainLogger.get());
	return gMainLogger;
}



int main(int argc, char* argv[])
{
    auto shared_logger = init(argv[0]);

#if _DEBUG
	std::cout << "Will crash";
#else
	std::cout << "Will hang";
#endif
	return 0;
}
```


Another example where you explicitly can reset it is shown as below
```
#include <iostream>
#include <g3log/g3log.hpp>
#include <g3log/logworker.hpp>
#include <g3log/std2_make_unique.hpp>


namespace {
	std::shared_ptr<g3::LogWorker> gMainLogger;
}

void init(std::string prefix)
{
	 gMainLogger.reset(g3::LogWorker::createLogWorker().release());

	auto defaultHandler = gMainLogger->addSink(std2::make_unique<g3::FileSink>(prefix, "./"),
		&g3::FileSink::fileWrite);
	g3::initializeLogging(gMainLogger.get());
}



int main(int argc, char* argv[])
{
    init(argv[0]);

#if _DEBUG
	std::cout << "Will crash";
#else
	std::cout << "Will hang";
#endif

	gMainLogger.reset();
	return 0;
}
```

The trick here is for you to make sure that your exitInstance function have access to the shared_ptr

**Comment by cstamatopoulos** [comment]
First option is not going to work for me as I have not access to main so when the shared_ptr goes out of scope the logging is going to shut down.

However, second option works like a charm. 

Thanks for your help, much appreciated.

**Comment by KjellKod** [comment]
Glad it worked :)

The first example only used main as an example. A better use case would be to then set the smart pointer to somewhere where the RAII would be triggered at shutdown.,,, or using it as in the 2nd example
